cpp/src/structs/copying/concatenate.cu

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/concatenate.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/concatenate.cuh>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/structs/structs_column_view.hpp>
+
+#include <structs/utilities.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace structs {
+namespace detail {
+
+/**
+ * @copydoc cudf::structs::detail::concatenate
+ *
+ */
+std::unique_ptr<column> concatenate(
+  std::vector<column_view> const& columns,
+  cudaStream_t stream                 = 0,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  // get ordered children
+  auto ordered_children = extract_ordered_struct_children(columns);
+
+  // concatenate them
+  std::vector<std::unique_ptr<column>> children;
+  children.reserve(columns[0].num_children());
+  std::transform(ordered_children.begin(),
+                 ordered_children.end(),
+                 std::back_inserter(children),
+                 [mr, stream](std::vector<column_view> const& cols) {
+                   return cudf::detail::concatenate(cols, mr, stream);
+                 });
```

Comment:
I think you can use a `thrust::transform_iterator` here and add `const` to the children vector.
```suggestion
auto begin = thrust::make_transform_iterator(ordered_children.begin(),
                                             ordered_children.end(),
                                             [mr, stream](std::vector<column_view> const& cols) {
                                               return cudf::detail::concatenate(cols, mr, stream);
                                             });
                                             
auto const children =
  std::vector<std::unique_ptr<column>>(begin, begin + columns[0].num_children());
```

Reply:
Can't do this because we do a std::move(children) later on.